### PR TITLE
Fixes #15451: incompletenesses in guard condition regarding cofix and primitive projections

### LINKED ISSUE #15451: Incompletenesses of guard condition in the presence of cofix redexes or not enough immediately applied primitive projection
#### Description of the problem

Here are two examples of fixpoints whose normal form is guarded but an unreduced cofix redex in the way blocks the normalization.

```coq
CoInductive Stream := {hd : nat; tl : Stream}.
Definition zeros := cofix zeros := {|hd := 0; tl := zeros|}.

Fixpoint f n :=
  match n with
  | 0 => 0
  | S n =>
    match zeros with
    | {|hd:=_|} => fun f => f n
    end f
  end.
```

```coq
Set Primitive Projections.
CoInductive Stream A := {hd : A; tl : Stream A}.
Arguments hd {A} s.

Fixpoint f n :=
  match n with
  | 0 => 0
  | S n => (cofix cst := {|hd := (fun f => f n); tl := cst|}).(hd) f
  end.
```

Found while working on #15434.

#### Coq Version

At least since Coq 8.3 (and probably 8.0) for the former, 8.5 for the latter, with #9602 in 8.10 having fixed most of preexisting incompletenesses.

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** Fix

Two issues:
- the case of a `cofix` was missing when detecting the presence of a potential iota-redex to reduce
- the stack was discarded when reducing a primitive projection

Fixes #15451

- [x] Added / updated **test-suite**.
- [x] Added **changelog**.

### CODE REVIEW

File: kernel/inductive.ml [hunk 1]
```diff
@@ -1209,8 +1209,8 @@ let check_one_fix renv recpos trees def =
               let c = whd_all renv.env c in
               let hd, _ = decompose_app c in
               match kind hd with
-              | Construct _ ->
-                  check_rec_call renv []
+              | Construct _ | CoFix _ ->
+                  check_rec_call renv stack
                     (Term.applist (mkProj(Projection.unfold p,c), l))
               | _ -> Exninfo.iraise exn
```
Comment (Alizter) [inline kernel/inductive.ml:1215]: Why not make this explicit so we definitely don't forget any cases?

Comment (herbelin) [reply kernel/inductive.ml:1215]: Good question. We could but there are a lot of cases to add. Wondering what others would do themselves.

Comment (ppedrot) [reply kernel/inductive.ml:1215]: I think that in the kernel at least, it is considered good practice to fully enumerate the constructors of `constr`.

Comment (mattam82) [reply kernel/inductive.ml:1215]: I agree, the catch-all bit me once in the guard as well when we introduced primitive projections.

Comment (herbelin) [reply kernel/inductive.ml]: OK, done.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
This one is fairly simple and there is little risk of anything going wrong, so let me assign and merge.

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**